transport/src/main/java/io/netty/channel/FileChannelBasedSlicedFileRegion.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2017 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.channel;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.WritableByteChannel;
+
+/**
+ * A derived {@link FileChannelBasedFileRegion} which shares reference counting with its parent.
+ */
+final class FileChannelBasedSlicedFileRegion extends FileChannelBasedFileRegion {
+
+    private final FileChannelBasedFileRegion parent;
+
+    private final long offset;
+
+    private final long count;
+
+    public FileChannelBasedSlicedFileRegion(FileChannelBasedFileRegion parent, long index,
+            long count) {
+        if (index < 0 || index > parent.count() - count) {
+            throw new IndexOutOfBoundsException(parent + ".slice(" + index + ", " + count + ')');
+        }
+
+        if (parent instanceof FileChannelBasedSlicedFileRegion) {
+            FileChannelBasedSlicedFileRegion sliced = (FileChannelBasedSlicedFileRegion) parent;
+            this.parent = sliced.parent;
+            this.offset = sliced.offset + index;
+        } else {
+            this.parent = parent;
+            this.offset = index;
+        }
+        this.count = count;
+    }
+
+    @Override
+    public long position() {
+        return parent.position() + offset;
+    }
+
+    @Override
+    public long transfered() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long transferred() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public long count() {
+        return count;
+    }
+
+    @Override
+    public long transferTo(WritableByteChannel target, long position) throws IOException {
+        throw new UnsupportedOperationException();
```

Comment:
> it makes it a bit difficult to map the behavior to the new index based solution.

I'd generally expect the two APIs to be independent: doing calls on one of the APIs doesn't update the indexes used by the other. I could even accept the two APIs being mutually exclusive: doing calls on one of the APIs breaks the other API (e.g., it might throw an exception).

> When given a FileRegion then you can use the old API, and for SliceableFileRegion, you'd better use the new API but it is still OK to use the old API, but it is not allowed to use the old API at first and then switch to the new API, and vice versa.

Yeah, that sounds fine. I'm not sure I'd go as far as saying it's "not allowed" to mix the APIs, but simply "it makes no sense". The result is the same either way, but the reasoning is a bit more clear.

That's why I think you could use this code, maybe as-is:
```java
private long transferred;

public long transferred() { return transferred; }
￼
public long transferTo(WritableByteChannel target, long position) throws IOException {
  long written = transferBytesTo(target, position, this.count - position);
  if (written > 0) {
    this.transferred += written;
  }
  return written;
}
```